radio/src/gui/colorlcd/draw_functions.cpp

```diff
@@ -487,3 +487,190 @@ void drawHexNumber(BitmapBuffer * dc, coord_t x, coord_t y, uint32_t val, LcdFla
+  coord_t y;
+
+  for (y = y0; y <= last; y++) {
+    a = x0 + sa / dy01;
+    b = x0 + sb / dy02;
+    sa += dx01;
+    sb += dx02;
+    if (a > b) SWAP(a, b);
+    dc->drawSolidHorizontalLine(a, y, b - a + 1, flags);
+  }
+
+  sa = dx12 * (y - y1);
+  sb = dx02 * (y - y0);
+
+  for (; y <= y2; y++) {
+    a = x1 + sa / dy12;
+    b = x0 + sb / dy02;
+    sa += dx12;
+    sb += dx02;
+    if (a > b) SWAP(a, b);
+    dc->drawSolidHorizontalLine(a, y, b - a + 1, flags);
+  }
+}
+
+
+// Cohen-Sutherland, see wikipedia
+// Compute the bit code for a point (x, y) using the clip rectangle
+// bounded diagonally by (xmin, ymin), and (xmax, ymax)
+uint8_t ComputeOutCode(float x, float y, float xmin, float xmax, float ymin, float ymax)
+{
+  uint8_t code = 0;
+  if (x < xmin)
+    code |= 1;
+  else if (x > xmax)
+    code |= 2;
+  if (y < ymin)
+    code |= 8;
+  else if (y > ymax)
+    code |= 4;
+  return code;
+}
+
+// clipping algorithm clips a line from P0 = (x0, y0) to P1 = (x1, y1) against a rectangle with
+// diagonal from (xmin, ymin) to (xmax, ymax).
+void CohenSutherlandLineClipAndDraw(BitmapBuffer * dc, float x0, float y0, float x1, float y1, float xmin, float xmax, float ymin, float ymax, uint8_t pat, LcdFlags flags)
+{
+  uint8_t outcode0 = ComputeOutCode(x0, y0, xmin, xmax, ymin, ymax);
+  uint8_t outcode1 = ComputeOutCode(x1, y1, xmin, xmax, ymin, ymax);
+  bool accept = false;
+
+  while (true) {
+    if (!(outcode0 | outcode1)) {
+      accept = true;
+      break;
+    }
+    else if (outcode0 & outcode1) {
+      break;
+    }
+    else {
+      float x = 0.0f, y = 0.0f;
+      uint8_t outcode = (outcode1 > outcode0) ? outcode1 : outcode0;
+      if (outcode & 4) {
+        x = x0 + (x1 - x0) * (ymax - y0) / (y1 - y0);
+        y = ymax;
+      }
+      else if (outcode & 8) {
+        x = x0 + (x1 - x0) * (ymin - y0) / (y1 - y0);
+        y = ymin;
+      }
+      else if (outcode & 2) {
+        y = y0 + (y1 - y0) * (xmax - x0) / (x1 - x0);
+        x = xmax;
+      }
+      else if (outcode & 1) {
+        y = y0 + (y1 - y0) * (xmin - x0) / (x1 - x0);
+        x = xmin;
+      }
+      if (outcode == outcode0) {
+        x0 = x;
+        y0 = y;
+        outcode0 = ComputeOutCode(x0, y0, xmin, xmax, ymin, ymax);
+      }
+      else {
+        x1 = x;
+        y1 = y;
+        outcode1 = ComputeOutCode(x1, y1, xmin, xmax, ymin, ymax);
+      }
+    }
+  }
+  if (accept) {
+    dc->drawLine(coord_t(x0+0.5f), coord_t(y0+0.5f), coord_t(x1+0.5f), coord_t(y1+0.5f), pat, flags);
+  }
+}
+
+void drawLineWithClipping(BitmapBuffer * dc, coord_t x0, coord_t y0, coord_t x1, coord_t y1, coord_t xmin, coord_t xmax, coord_t ymin, coord_t ymax, uint8_t pat, LcdFlags flags)
+{
+  CohenSutherlandLineClipAndDraw(dc, x0, y0, x1, y1, xmin, xmax, ymin, ymax, pat, flags);
+}
+
+void drawHudRectangle(BitmapBuffer * dc, float pitch, float roll, coord_t xmin, coord_t xmax, coord_t ymin, coord_t ymax, LcdFlags flags)
```

Comment:
All the primitives used in this function are already implemented. The question is why should be have a HUD drawing function in the generic code? This is only required for very special apps and can be easily included in LUA.

Reply:
where is a HUD drawing function in the generic code?

> This is only required for very special apps and can be easily included in LUA.

I very strongly disagree

I have this suspicion for a while now, from the discussion in this repo and the discord, that you guys tend to forget that our transmitters are quite resource constraint platforms and disregard performance (As well es memory) considerations way too often :(

having this HUD implemented in the C code and not lua MASSIVELY speeds up the lua script execution !!! It's actually all reported and documented elsewhere. E.g., for my script I found that the scrip execution had doubled (= DOUBLED!!!) just by replacing the lua HUD with the C code HUD ...

so, frankly, and I'm sorry to say that so clearly and unambiguously, when I hear some of you saying "can be easily included in LUA" my hairs stand up ... you must be living on a different planet than me :D

Reply:
EDIT: I'd like to add that at least from my subjective counting the apps which are used by far most do have a HUD ...

Reply:
Then please place that HUD function at least directly in the LUA API file for color lcd.

Reply:
??? I really strungle sometimes with following your logic

I have put it specifically in the draw_functions files since it also holds drawTimer(), drawSensorCustomValue(), drawSwitch(), drawSource(), all used in api_colorlcd, and I exactly followed that pattern ... so either all or none I would say :)

Reply:
> and I exactly followed that pattern ... so either all or none I would say :)

As it is used only by LUA functions, I prefer to have this code directly with the other LUA API functions. The other functions you are referring to are / were used at other places as well.